crates/matrix-sdk-ui/src/timeline/pagination.rs

```diff
@@ -12,9 +12,108 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use std::{fmt, ops::ControlFlow};
+use std::{fmt, ops::ControlFlow, pin::pin, sync::Arc, time::Duration};
+
+use matrix_sdk::{room::MessagesOptions, Result};
+use matrix_sdk_base::timeout::timeout;
+use ruma::assign;
+use tracing::{debug, error, info, warn};
+
+use super::{inner::HandleBackPaginatedEventsError, Timeline};
+
+impl Timeline {
+    pub(super) async fn paginate_backwards_impl(
+        &self,
+        mut options: PaginationOptions<'_>,
+    ) -> Result<ControlFlow<BackPaginationStatus>> {
+        let mut from = match self.inner.back_pagination_token().await {
+            None if options.wait_for_token => {
+                let notified = pin!(self.sync_response_notify.notified());
+                match timeout(notified, Duration::from_secs(3)).await {
+                    Ok(()) => match self.inner.back_pagination_token().await {
+                        Some(token) => Some(token),
+                        None => {
+                            warn!(
+                                "Sync response without prev_batch received, \
+                                 start of timeline reached?"
+                            );
+                            return Ok(ControlFlow::Break(BackPaginationStatus::Idle));
+                        }
+                    },
+                    Err(_) => {
+                        debug!("Waiting for prev_batch token timed out after 3s");
+                        None
+                    }
+                }
+            }
+            token => token,
+        };
+
+        self.back_pagination_status.set_if_not_eq(BackPaginationStatus::Paginating);
+
+        let mut outcome = PaginationOutcome::new();
+        while let Some(limit) = options.next_event_limit(outcome) {
+            let messages = self
+                .room()
+                .messages(assign!(MessagesOptions::backward(), {
+                    from: from.clone(),
+                    limit: limit.into(),
+                }))
+                .await?;
+            let chunk_len = messages.chunk.len();
+
+            let tokens = PaginationTokens { from, to: messages.end.clone() };
+            let res = match self.inner.handle_back_paginated_events(messages.chunk, tokens).await {
+                Ok(result) => Some(result),
+                Err(HandleBackPaginatedEventsError::TokenMismatch) => {
+                    info!("Start of timeline was altered since pagination was started, resetting");
+                    return Ok(ControlFlow::Continue(()));
+                }
+                Err(HandleBackPaginatedEventsError::ResultOverflow) => None,
+            };
+
+            let process_events_result = async {
```

Comment:
nit: spurious async here? I don't see any await in this async block :eyes:

Reply:
Ah yeah, this would ideally be a try block now. I guess it makes sense to change to to an immediately evaluated closure.